my/core/query.py

```diff
@@ -0,0 +1,717 @@
+    m = _mixed_iter()
+    for _ in range(3):
+        yield next(m)
+    yield RuntimeError("Unhandled error!")
+    yield from m
+
+
+def test_order_value() -> None:
+
+    default_order = list(_mixed_iter())
+
+    # if the value for some attribute on this item is a datetime
+    sorted_by_datetime = list(select(_mixed_iter(), order_value=lambda o: isinstance(o, datetime)))
+    assert len(sorted_by_datetime) == 6
+
+    mixed_iter_sorted_indexes = (4,5,2,3,0,1)
+
+    for val, expected_index in zip(sorted_by_datetime, mixed_iter_sorted_indexes):
+        assert val == default_order[expected_index]
+
+
+def test_key_clash() -> None:
+
+    import pytest
+
+    # clashing keys causes errors if you use order_key
+    with pytest.raises(TypeError, match=r"not supported between instances of 'datetime.datetime' and 'int'") as te:
+        list(select(_mixed_iter(), order_key="y"))
+
+
+def test_wrap_unsortable() -> None:
+
+    from collections import Counter
+
+    # by default, wrap unsortable
+    res = list(select(_mixed_iter(), order_key="z"))
+    assert Counter(map(lambda t: type(t).__name__, res)) == Counter({"_A": 4, "Unsortable": 2})
+
+
+def test_disabled_wrap_unsorted() -> None:
+
+    import pytest
+
+    # if disabled manually, should raise error
+    with pytest.raises(TypeError, match=r"not supported between instances of 'NoneType' and 'int'") as te2:
+        list(select(_mixed_iter(), order_key="z", wrap_unsorted=False))
+
+
+def test_drop_unsorted() -> None:
+
+    from collections import Counter
+
+    # test drop unsortable, should remove them before the 'sorted' call
+    res = list(select(_mixed_iter(), order_key="z", wrap_unsorted=False, drop_unsorted=True))
+    assert len(res) == 4
+    assert Counter(map(lambda t: type(t).__name__, res)) == Counter({"_A": 4})
+
+
+def test_drop_exceptions() -> None:
+
+    assert more_itertools.ilen(_mixed_iter_errors()) == 7
+
+    # drop exceptions
+    res = list(select(_mixed_iter_errors(), order_value=lambda o: isinstance(o, datetime), drop_exceptions=True))
+    assert len(res) == 6
+
+
+def test_raise_exceptions() -> None:
+
+    import pytest
+
+    # raise exceptions
+    with pytest.raises(RuntimeError) as r:
+        select(_mixed_iter_errors(), order_value=lambda o: isinstance(o, datetime), raise_exceptions=True)
+    assert str(r.value) == "Unhandled error!"
+
+
+def test_wrap_unsortable_with_error_and_warning() -> None:
+
+    import pytest
+    from collections import Counter
+
+    # by default should wrap unsortable (error)
+    with pytest.warns(UserWarning, match=r"encountered exception") as w:
+        res = list(select(_mixed_iter_errors(), order_value=lambda o: isinstance(o, datetime)))
+    assert Counter(map(lambda t: type(t).__name__, res)) == Counter({"_A": 4, "_B": 2, "Unsortable": 1})
+    # compare the returned error wrapped in the Unsortable
+    returned_error = next(filter(lambda o: isinstance(o, Unsortable), res)).obj  # type: ignore[union-attr]
```

Comment:
not super important, but I think the ignore is needed here because mypy can't peek into the lambda. I usually use something like `(o for o in res if isinstance(r, Exception))` -- same length and mypy can guess the type correctly!